Autogen webgpu dawn, removing wgpu-py dependency (f16 support part 1)

wpmed92 opened the pull request:
This PR changes the WebGPU runtime used from `wgpu` to `Dawn` (used by Chrome):
- `wgpu` is not closely following the WebGPU specs, it doesn't impose a limit on storage buffers per kernel, but Dawn does, so some tests had to be skipped due to more than the supported number of buffers per kernel
- the python bindings used are created by me, it's published as `dawn-python` (supports both Mac (Metal), and Linux (Vulkan)
- WebGPU specs doesn’t allow using the same buffer for input and output, JIT does this, and was fine on wgpu (non spec compliant). I patch it in ops_webgpu

Here is a demo of `f16` stable diffusion deployed from another branch that includes the `f16` support: https://wpmed92.github.io/stable-diffusion-tinygrad-f16/
I see around 30% perf gain from `f16` in SD.
The `f16` `wgsl` change is coming in a follow-up PR, I didn't want to put these two relatively big changes into 1 PR.

geohot commented:
Yea, I'm fine with grabbing the bins from a GitHub release (like remu or ocelot)

geohot commented:
Progress on adding dawn bindings to autogen?

wpmed92 commented:
@geohot I moved dawn-python stuff to tinygrad autogen, I have all tests passing locally on my Mac, but have to clean it up and make it pass in CI.

geohot commented:
I don't think there should be a support for this. Why can't it go in the webgpu file?

wpmed92 commented:
@geohot 
- Moved removed webgpu support and moved it to ops_webgpu
- refactored callback info creation -> removed a lot of the duplications
- Fixed autogen verification
- Fixed mypy

Everything is passing, except for the line limit test.
I'm 100 lines above the limit, but there's more to save in ops_webgpu than what i saved so far, so will work on it more.

wpmed92 commented:
From the inital 421 lines ops_webgpu it's down to 324, need to save only 28 more.

<img width="358" alt="Screenshot 2025-01-29 at 15 03 24" src="https://github.com/user-attachments/assets/e709d096-2c79-4ddf-ab25-84f7dea2c615" />

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,323 @@
-    command_encoder = self.dev.create_command_encoder()
+    bindings += [{"binding": i+1, "resource": {"buffer": create_uniform(self.dev, x) if i >= len(tmp_bufs) else x, "offset": 0,
+                                            "size": 4 if i >= len(tmp_bufs) else webgpu.wgpuBufferGetSize(x)}} for i,x in enumerate(tuple(tmp_bufs)+vals)]  # noqa: E501
+    bind_group_layout = create_bind_group_layout(self.dev, entries=binding_layouts)
```
Why are all these things functions if they are only called in one place?

wpmed92 commented:
`36` lines saved with removing single-use functions plus other refactors: `11028 -> 10992`

wpmed92 commented:
`44` more lines saved with more refactors. Every single use function is deleted now and are used inline.
I was below 11k, but after rebasing I'm at 11027.

rcastill commented:
Hi, I tried replicating this by running the yolov8 webgpu example.

First I downloaded the lib the same way as declared in github actions:

```
sudo curl -L https://github.com/wpmed92/pydawn/releases/download/v0.1.6/libwebgpu_dawn.so -o /usr/local/lib/libwebgpu_dawn.so
```

I had this problem:

```
OSError: libdawn.so: cannot open shared object file: No such file or directory
```

Solved with `sudo ldconfig`. That essentially made a dynamic link `/usr/local/libdawn.so -> /usr/local/libwebgpu_dawn.so`.

Now I'm stuck at:

```
    raise RuntimeError(f"Failed to request device: [{webgpu.WGPURequestDeviceStatus__enumvalues[device_result[0]]}] {device_result[2]}")
RuntimeError: Failed to request device: [WGPURequestDeviceStatus_Error] Invalid feature required: Requested feature FeatureName::ShaderF16 is not supported.
    at CreateDeviceInternal (/__w/webgpu-dawn-binaries/webgpu-dawn-binaries/build/_deps/dawn-src/src/dawn/native/Adapter.cpp:326)
```

Am I missing something @wpmed92?

wpmed92 commented:
Looks like shader_f16 is not supported on your device. It's not available on all machines, just like timestamp queries.

rcastill commented:
That's odd. I have a nvidia gtx1080. I've compiled other projects with half precision enabled.

Maybe it's something specific to dawn for my device?

wpmed92 commented:
Yeah, I didn't mean your device doesn't support. Dawn doesn't support it on your device.

wpmed92 commented:
@geohot Tests all green now.

geohot commented:
On my Mac
```
git checkout -b wpmed92-switch-wgpu-to-dawn master
git pull https://github.com/wpmed92/tinygrad.git switch-wgpu-to-dawn
WEBGPU=1 python3 test/test_tiny.py
```

```
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/Users/light/tinygrad/tinygrad/runtime/ops_webgpu.py", line 11, in <module>
    instance = webgpu.wgpuCreateInstance(instDesc)
               ^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'tinygrad.runtime.autogen.webgpu' has no attribute 'wgpuCreateInstance'
```

geohot commented:
New issue

`WEBGPU=1 python3 examples/beautiful_mnist.py`

```
...
  File "/Users/light/tinygrad/tinygrad/codegen/lowerer.py", line 68, in get_index
    idxs = get_grouped_dims("gidx", full_shape[:global_dims], opts.global_max, reverse=True) + \
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/light/tinygrad/tinygrad/codegen/lowerer.py", line 32, in get_grouped_dims
    limited = _limit_dims(dims, max_sizes) if max_sizes is not None else dims
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/light/tinygrad/tinygrad/codegen/lowerer.py", line 24, in _limit_dims
    if dims[i] * dims[i+1] <= m:
                 ~~~~^^^^^
IndexError: tuple index out of range
```

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,280 @@
+    toggle_desc = webgpu.WGPUDawnTogglesDescriptor()
+    toggle_desc.chain.sType = webgpu.WGPUSType_DawnTogglesDescriptor
+    toggle_desc.enabledToggleCount = 1
+    unsafe_apis = ctypes.cast(ctypes.pointer(to_c_string("allow_unsafe_apis")), ctypes.POINTER(ctypes.c_char))
```
Hmm, I feel like some of this stuff can be made more readable and shorter with helpers

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,280 @@
+    pipeline_layout_desc = webgpu.WGPUPipelineLayoutDescriptor()
+    pipeline_layout_desc.bindGroupLayoutCount = len(bind_group_layouts)
+    bind_group_array_type = webgpu.WGPUBindGroupLayout * len(bind_group_layouts)
+    pipeline_layout_desc.bindGroupLayouts = bind_group_array_type(*bind_group_layouts)
```
A nit, but I think it's more readable like this

`pipeline_layout_desc.bindGroupLayouts = (webgpu.WGPUBindGroupLayout * len(bind_group_layouts))(*bind_group_layouts)`

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,280 @@
+
+    # Check compiler error
+    webgpu.wgpuDevicePushErrorScope(self.dev, webgpu.WGPUErrorFilter_Validation)
+    shader_module = webgpu.wgpuDeviceCreateShaderModule(self.dev, module)
```
Not required for this PR, but is there a way for the Compiler to be separate from the Program?

geohot commented:
Close to ready, three main points:
* I think there's at least 10-20 lines to save making things both shorter and more readable. This is a big positive diff to lines
* If the library is missing, it should output instructions to fix. We should be better about this everywhere
* beautiful_mnist is broken, I expect all the basic examples to work on Mac.

geohot commented:
I'll also raise this bounty to $400, I think the removal of wgpu-py has value beyond just float16

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
+
+def read_buffer(dev, buf):
+  size = webgpu.wgpuBufferGetSize(buf)
+  tmp_usage = webgpu.WGPUBufferUsage_CopyDst | webgpu.WGPUBufferUsage_MapRead
```
Why is this a variable?

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
+
+def write_buffer(device, buf, offset, src):
+  src = bytearray(src)
+  src_pointer = (ctypes.c_uint8 * len(src)).from_buffer(src)
```
Why is this a variable?

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
-    buffer_data = self.dev.queue.read_buffer(src, 0)
-    dest[:] = buffer_data[:dest.nbytes] if src._nbytes > dest.nbytes else buffer_data
+    buffer_data = read_buffer(self.dev, src)
+    src_len = webgpu.wgpuBufferGetSize(src)
```
Why is this a variable?

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
   def _alloc(self, size: int, options):
-    return self.dev.create_buffer(size=round_up(size, 4), usage=wgpu.BufferUsage.STORAGE | wgpu.BufferUsage.COPY_DST | wgpu.BufferUsage.COPY_SRC)
+    # WebGPU buffers have to be 4-byte aligned
+    return webgpu.wgpuDeviceCreateBuffer(
```
Why does this have so many lines?

wpmed92 replied:
I made it 1-line less, but because of verbose webgpu method names it won't fit 1 line as before.

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
+
+    if wait: webgpu.wgpuCommandEncoderResolveQuerySet(command_encoder, query_set, 0, 2, query_buf, 0)
+
+    cmd_buf = webgpu.wgpuCommandEncoderFinish(command_encoder, webgpu.WGPUCommandBufferDescriptor())
```
Why is this a variable?

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
+
+    def cb(status, compute_pipeline_impl, msg, u1, u2): pipeline_result[:] = status, compute_pipeline_impl, from_wgpu_str(msg)
+
+    cb_info = create_cb_info(webgpu.WGPUCreateComputePipelineAsyncCallbackInfo2,  webgpu.WGPUCreateComputePipelineAsyncCallback2, cb)
```
Why is this a variable? (multiple places)

geohot commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -2,62 +2,252 @@
+    binding_layouts += [webgpu.WGPUBindGroupLayoutEntry(binding=i+1, visibility=webgpu.WGPUShaderStage_Compute,
+      buffer=webgpu.WGPUBufferBindingLayout(type=webgpu.WGPUBufferBindingType_Uniform if i >= len(tmp_bufs)
+      else webgpu.WGPUBufferBindingType_Storage)) for i in range(len(tmp_bufs)+len(vals))]
+    bindings = [webgpu.WGPUBindGroupEntry(binding=0, buffer=create_uniform(self.dev, float('inf')), offset=0, size=4)]
```
Construct this where it's used?

geohot commented:
Bumped the line limit to merge this, really happy to have that wgpu dep gone.

Korijn commented:
> Bumped the line limit to merge this, really happy to have that wgpu dep gone.

Hello, just checking in here as one of the authors of wgpu-py, would any of you mind summarizing your reasons to move away from our library? Anything we could do better? Thanks!

cc @almarklein

wpmed92 commented:
@Korijn 
The reason for the removal has nothing to do with something wgpu-py did wrong. I love wgpu-py and used it as a reference when rewriting our runtime with Dawn (see read_buffer implementation as an example).
The removal has to do with the general direction of the project and the tinygrad philosophy: reducing complexity, and removing dependencies. 
We try to not use python interface libs, i.e. libs that provide a wrapper interface around a binary. Instead, we autogen this interface with clang2py, and we only depend on the binary (Dawn in this case).
We also removed llvmlite and use autogened llvm instead. We also use autogened opencl, etc.
As to what could be better, but it’s still not wgu-py: I think that maybe wgpu (not wgpu-py) is moving a bit slowly, f16 is still not merged although it’s close. (the pr is open since last may)
